patches/0005-Add-crypto-backend-foundation.patch

```diff
@@ -0,0 +1,322 @@
+
+diff --git a/src/crypto/internal/backend/backend_test.go b/src/crypto/internal/backend/backend_test.go
+new file mode 100644
+index 00000000000..c2c06d3bff8
+--- /dev/null
++++ b/src/crypto/internal/backend/backend_test.go
+@@ -0,0 +1,30 @@
++// Copyright 2017 The Go Authors. All rights reserved.
++// Use of this source code is governed by a BSD-style
++// license that can be found in the LICENSE file.
++
++package backend
++
++import (
++	"testing"
++)
++
++// Test that Unreachable panics.
++func TestUnreachable(t *testing.T) {
++	defer func() {
++		if Enabled {
++			if err := recover(); err == nil {
++				t.Fatal("expected Unreachable to panic")
++			}
++		} else {
++			if err := recover(); err != nil {
++				t.Fatalf("expected Unreachable to be a no-op")
++			}
++		}
++	}()
++	Unreachable()
++}
++
++// Test that UnreachableExceptTests does not panic (this is a test).
++func TestUnreachableExceptTests(t *testing.T) {
++	UnreachableExceptTests()
++}
+diff --git a/src/crypto/internal/backend/bbig/big.go b/src/crypto/internal/backend/bbig/big.go
+new file mode 100644
+index 00000000000..51bc3c68048
+--- /dev/null
++++ b/src/crypto/internal/backend/bbig/big.go
+@@ -0,0 +1,17 @@
++// Copyright 2022 The Go Authors. All rights reserved.
++// Use of this source code is governed by a BSD-style
++// license that can be found in the LICENSE file.
++
++//go:build !boringcrypto && !goexperiment.opensslcrypto
++
++package bbig
++
++import "math/big"
++
++func Enc(b *big.Int) []uint {
++	return nil
++}
++
++func Dec(b []uint) *big.Int {
++	return nil
++}
+diff --git a/src/crypto/internal/backend/common.go b/src/crypto/internal/backend/common.go
+new file mode 100644
+index 00000000000..672433ea880
+--- /dev/null
++++ b/src/crypto/internal/backend/common.go
+@@ -0,0 +1,65 @@
++// Copyright 2022 The Go Authors. All rights reserved.
++// Use of this source code is governed by a BSD-style
++// license that can be found in the LICENSE file.
++
++package backend
++
++import (
++	"crypto/internal/boring/sig"
++	"runtime"
++	"syscall"
++)
++
++func init() {
++	if v, _ := envGoFIPS(); v == "1" {
++		if runtime.GOOS != "linux" {
++			panic("FIPS mode requested (GOFIPS=1) but not supported on " + runtime.GOOS)
++		} else if !iscgo {
++			panic("FIPS mode requested (GOFIPS=1) but not supported with cgo disabled")
++		}
++	}
++}
++
++func envGoFIPS() (string, bool) {
++	if v, ok := syscall.Getenv("GOFIPS"); ok {
++		return v, true
++	}
++	// TODO: Decide which environment variable to use.
++	// See https://github.com/microsoft/go/issues/397.
++	if v, ok := syscall.Getenv("GOLANG_FIPS"); ok {
++		return v, true
++	}
++	return "", false
++}
++
++// Unreachable marks code that should be unreachable
++// when backend is in use.
++func Unreachable() {
++	if Enabled {
++		panic("cryptobackend: invalid code execution")
++	} else {
++		// Code that's unreachable is exactly the code
++		// we want to detect for reporting standard Go crypto.
++		sig.StandardCrypto()
++	}
++}
```

Comment:
Your branch elimination and the linker work like you described. On the other hand, `go version` does not report the existence of `sig.*()` functions, but it does report `GOEXPERIMENTS`.

I've tested your assumptions about `rsc/goversion`, which does understand `sig.*()` functions, and these are the results:
- `GOEXPRIMENT=opensslcrypto`: Only standard crypto is reported. It does not detect `sig.BoringCrypto()` because the function name created by the linker has changed from `crypto/internal/boring/sig.BoringCrypto` to `crypto/internal/boring/sig.BoringCrypto.abi0`, and `goversion` does an strict comparison against the former (see [code](https://github.com/rsc/goversion/blob/597212e462da05a7902d6cea0ec895a0d9b8b218/version/read.go#L60)).
- `GOEXPERIMENT=boringcrypto`: Reports both standard and boring crypto. It neither finds `sig.BoringCrypto` due to previous reasoning, but it does find functions starting with `_Cfunc__goboringcrypto_`, which has the same effect. Standard crypto should not be reported because boring is a compile time switch and `sig.StandardCrypto` does not appear in the final binary, but it also looks for [these function names](https://github.com/rsc/goversion/blob/597212e462da05a7902d6cea0ec895a0d9b8b218/version/read.go#L111), and the `sha1` can't branch-eliminate `sha1.digest` due to [this change](https://github.com/microsoft/go/blob/86b601f5fffce395dee5c319017d4e7c029e4f02/patches/0006-Integrate-OpenSSL-module.patch#L408) in our patches.

Reply:
I've submitted https://go-review.googlesource.com/c/go/+/413754/ upstream so we don't have to redefine `sha1.boringEnabled` as a variable instead of a constant.

I would rather keep `borincrypto` linking `sha1.digest` until upstream accepts my changes, as fixing it on our side would mean more diffs to maintain for a backend that we don't really officially support.

Reply:
> * `GOEXPRIMENT=opensslcrypto`: Only standard crypto is reported. It does not detect `sig.BoringCrypto()` because the function name created by the linker has changed from `crypto/internal/boring/sig.BoringCrypto` to `crypto/internal/boring/sig.BoringCrypto.abi0`, and `goversion` does an strict comparison against the former (see [code](https://github.com/rsc/goversion/blob/597212e462da05a7902d6cea0ec895a0d9b8b218/version/read.go#L60) and [With golang 1.17 the symbols are different  rsc/goversion#20](https://github.com/rsc/goversion/issues/20)).

Interesting... seems strange at first glance that there's code to check for the sig byte sequences in `findCryptoSigs`, but that func doesn't end up being used if the code's able to find `runtime.buildVersion` (and a debug flag isn't set). And if it does end up using `findCryptoSigs`, it's *in addition to* the symbol name check.

Seems like there's probably some history here--one is more reliable over the other, or one works with different versions of Go than the other. Just my guess.

At the end of the day, `rsc.io/goversion` isn't part of the Go toolset, so I don't think we need to worry about it all that much. 🙂